blog: correct technical inaccuracies in the controller-runtime cache post (#56682)
kubernetes/website: content/en/blog/_posts/2026/controller-runtime-cache-explained/index.md

Reviewer: This will need a leading message - I recommend a https://www.docsy.dev/docs/content/shortcodes/#pageinfo - to say that the article has been revised, since initial publication, to address technical inaccuracies.

The pageinfo should come just after the front matter.

Author: Sure, done

  Timofei Larkin (Ænix)
---

{{% pageinfo color="primary" %}}
**Update:** *This article has been revised since it was first published, to correct technical

> inaccuracies in the original text.*

Reviewer: ```suggestion
This article has been revised since it was first published, to correct several significant technical inaccuracies in the original text.
```

Reviewer: It's important to be honest about how big the inaccuracies were.

Author: Applied as suggested.

happened after that version. That is the basis of consistency: there is no gap between the
snapshot and the stream, because the stream resumes exactly where the snapshot ended.

That snapshot no longer arrives as a separate **list** call by default. Current versions use a
[streaming list](/docs/reference/using-api/api-concepts/#streaming-lists) instead: the Reflector
opens the **watch** with `sendInitialEvents=true`, and the API server begins the stream with
synthetic `ADDED` events for the whole current state before switching to live changes. One
request instead of two, and a plain **list** as the fallback. The pattern is unchanged —
snapshot, then stream — which is why this article keeps saying **list** + **watch**, the way the
Kubernetes documentation does.

If the connection drops, the Reflector reconnects with the last known `resourceVersion`. If
the API server replies with `410 Gone` ("that version is no longer in the history, you are
too far behind"), the Reflector fetches a fresh snapshot and starts over. This is called a
_relist_, and it does not happen on a schedule — only in those failure scenarios.

> ### The delta queue

Reviewer 2: No request to change anything, but I find the in-depth explanation of delta fifo, reflector etc a bit odd, because the rest of the article is about how to correctly use controller-runtime and for that it doesn't really matter at all how client-go does watching under the hood

Author: That's a fair observation about genre. The article's stated premise is the internals tour itself: the audience is people who already use controller-runtime and want the mental model behind the advice. The usage guidance in the second half is meant to fall out of that model rather than stand alone.

server, "open a **watch** from version X", and receives a stream of events for everything that
happened after that version. That is the basis of consistency: there is no gap between the
snapshot and the stream, because the stream resumes exactly where the snapshot ended.

That snapshot no longer arrives as a separate **list** call by default. Current versions use a
[streaming list](/docs/reference/using-api/api-concepts/#streaming-lists) instead: the Reflector
opens the **watch** with `sendInitialEvents=true`, and the API server begins the stream with
synthetic `ADDED` events for the whole current state before switching to live changes. One
request instead of two, and a plain **list** as the fallback. The pattern is unchanged —
snapshot, then stream — which is why this article keeps saying **list** + **watch**, the way the
Kubernetes documentation does.

If the connection drops, the Reflector reconnects with the last known `resourceVersion`. If
the API server replies with `410 Gone` ("that version is no longer in the history, you are
too far behind"), the Reflector re-syncs: it fetches a fresh snapshot, first from the last

> version it observed, and only from scratch if that version is gone too. This is called a

Reviewer 3: > If the connection drops, the Reflector reconnects with the last known `resourceVersion`. If the API server replies with `410 Gone` [...]
> the Reflector re-syncs: it fetches a fresh snapshot, first from the last
version it observed

I think this paragraph states the same thing twice: the reflector connects with the last known resourceVersion, then it fetches a snapshot with the last observed version.
I presume, it would get a second 410 Gone when trying to fetch that snapshot.

So you get two layers with cleanly separated responsibilities:

> - **The delta queue** — an ordered stream of change facts, delivered one at a time and

Reviewer: ```suggestion
- The **delta queue** — an ordered stream of change facts, delivered one at a time and
```

Author: Applied.

  without merging. Its job is to tell consumers everything that happened, in the right order.

> - **workqueue** — a queue of **keys** with deduplication and rate limiting. This is the layer

Reviewer 2: I would not mention rate limiting here, because the ratelimiting is basically never used in the context of the handlers and its explicitly opt in

Author: Right, handlers enqueue with a plain Add; the rate limiter only comes into play on retries. Dropped it here.

`sync.RWMutex`, plus a dictionary of registered indexes (covered in their own section below).

An uncontended `r.Get` is cheap: a map lookup followed by a `DeepCopy` of the object. The
part of that structure that matters most at scale, though, is not the map — it is the one
`sync.RWMutex`, which guards the store and every index at once. Readers hold it for shared
access, the informer needs it exclusively to write, so the two genuinely compete: a `List`
holds the read lock while it walks every object of that kind and collects pointers — the deep
copies happen afterwards, outside the lock — and the next store write waits behind that walk.

That much is structural and still true. What it used to cause is not. In
`kube-controller-manager` at scale, full `List` calls held the read lock long enough to back the
informer queue up and let the cache go stale — see
[kubernetes#130767](https://github.com/kubernetes/kubernetes/issues/130767), fixed in the 1.36
cycle by making the writer take the lock far less often. The contention got much cheaper, but
the single lock is still there, and you get the improvements by compiling against a newer

> `client-go` rather than by upgrading the cluster.

Reviewer 3: By the time we get to issue numbers, we're no longer introducing a data structure for engineers trying to learn some concepts, instead we're turning this article into a changelog, build guidance and an incident postmortem. I think the entire last paragraph can be condensed into a footnote for the previous paragraph:

> [...] this was a real bottleneck in kube-controller-manager at scale (kubernetes#130767); recent client-go releases hold the write lock for much less time.

### SharedIndexInformer and subscriptions

A SharedIndexInformer fuses the Reflector, the delta queue, and the Indexer together and
exposes two interfaces to the rest of the world:

- Read objects directly from the indexer.
- Register a `ResourceEventHandler` and receive notifications for every event coming out of
  the queue — `OnAdd`, `OnUpdate`, `OnDelete`. The store is written before the handler call,
  so by the time your handler runs the indexer already reflects that event, and possibly
  later ones too.

"Outside" here means your controllers. When a controller registers `Watches(...)`, under the
hood it asks the informer: "add a handler that, on every change, enqueues the key into my
workqueue". The controller's workers then pop keys one at a time and call your
`Reconcile(ctx, ctrl.Request{NamespacedName: ...})`.

The keyword in the name is **Shared**. The manager creates **one** informer per GVK, and
every controller, webhook, and event source within that manager subscribes to it. (More
precisely, one per GVK *per flavour of object*: typed, `Unstructured`, and
`PartialObjectMetadata` are tracked separately, so reading Pods both as `corev1.Pod` and as

> metadata-only means two informers and two watches for `v1/Pod`.)

Reviewer 3: I think the parenthetical clarifications are not helping the purpose of the article. Technically correct, yes, but bloat the text with asides, digressions, and clarifications. This would be great as a footnote or expandable, which Hugo supports, but I've looked at some blog posts and grepped the website source and haven't found any use of such expressions, so not sure if it's doable in the k8s blog. Another alternative would be to move this clarification to the place where we introduce the metadata-only watch.

`r.Get` inside your reconciler:

1. The manager's `mgr.Start(ctx)` brings up every registered informer.

> 2. For each GVK, the Reflector fetches a full snapshot of every object that falls within your

Reviewer 2: Maybe to clarify this is about the list
```suggestion
2. For each GVK, the Reflector fetches a full snapshot of every object type that falls within your
```

Author: Took the spirit of the suggestion with one tweak: since the step is already per-GVK, went with "every object of that type" to avoid a type-within-type reading.

   scope.
3. The snapshot is loaded into the informer's store, registered indexes are rebuilt, and the

>    informer's `HasSynced()` flag flips to `true`.

Reviewer 2: I would probably not go so far as to spell out the method name and signature, in newer client-go this can be waited for in a channel rather than polling (although it appears c-r doesn't properly leverage that): https://github.com/kubernetes-sigs/controller-runtime/blob/b7090f65d5663ab151413ab07e6f2d1550e3963a/pkg/cache/cache.go#L132

Author: Done, dropped the method name.

4. The same stream then continues as an ordinary **watch** from the `resourceVersion` the
   snapshot synced to.

> 5. **Only then** does the controller start invoking `Reconcile` — specifically, once

Reviewer 2: It also waits for the handlers to finish processing the initial list (although I am not sure offhand what happens in the listwatch case): https://github.com/kubernetes-sigs/controller-runtime/blob/b7090f65d5663ab151413ab07e6f2d1550e3963a/pkg/internal/source/kind.go#L127

Author: Good point, folded in: the wait covers the handlers processing the initial snapshot, not just the store fill.

reconcile fixes it. You do not need to "wait 100ms" or "re-trigger". You need to write the
logic so that one or two extra invocations break nothing.

If a stale read is a genuine correctness problem for you, a live read does not fix it — a
concurrent write can be mid-commit anyway. See the

> [controller-runtime FAQ](https://github.com/kubernetes-sigs/controller-runtime/blob/main/FAQ.md#q-my-cache-might-be-stale-if-i-read-from-a-cache-how-should-i-deal-with-that)

Reviewer: nit: link to a stable ref, eg 820ed1a84f67c2f5c4ee4a104bc6c00faddb1384
```suggestion
[controller-runtime FAQ](https://github.com/kubernetes-sigs/controller-runtime/blob/820ed1a84f67c2f5c4ee4a104bc6c00faddb1384/FAQ.md#q-my-cache-might-be-stale-if-i-read-from-a-cache-how-should-i-deal-with-that)
```

It may go stale, but so will the article (we automatically add a "this is stale" banner to old blog articles).

Author: Pinned, and did the same for the link in the wrap-up.

to Pods.

Because Go has no immutable structs, nothing prevents you from doing
`pod.Labels["foo"] = "bar"` directly inside a handler — and that Pod is the one in the store.
Raw `client-go` listers have always worked this way; `ThreadSafeStore`'s own documentation puts
it bluntly: "you must not modify anything returned by Get or List as it will break the indexing

> feature". Patch a status "for convenience" in one controller and you break the world view of an

Reviewer 2: ```suggestion
feature". Patch a status "for convenience" in a handler and you break the world view of an
```

Author: Applied.

to Pods.

Because Go has no immutable structs, nothing prevents you from doing
`pod.Labels["foo"] = "bar"` directly inside a handler — and that Pod is the one in the store.
Raw `client-go` listers have always worked this way; `ThreadSafeStore`'s own documentation puts
it bluntly: "you must not modify anything returned by Get or List as it will break the indexing

> feature". Patch a status "for convenience" in a handler and you break the world view of an

Reviewer: ```suggestion
it bluntly: you must not modify anything returned by `Get` or `List` as it will break the indexing
feature. Patch a status "for convenience" in a handler and you break the world view of an
```

Author: Applied.

unrelated controller next door.

> `controller-runtime`'s cache-backed client shields you from that on the read path: `Get` and

Reviewer: ```suggestion
The cache-backed client from `controller-runtime` shields you from that on the read path: `Get` and
```

Author: Applied.

### Mistake 3: resync is not relist

An informer has a resync period (`cache.Options.SyncPeriod`, 10 hours by default in
`controller-runtime`), and many people read it as "rebuild the cache from the API server every
N hours".

It does not. A resync does **not** perform a **list**. It _re-emits_ everything currently in
the indexer back through the delta queue, and the informer dispatches an update per object,
calling `OnUpdate(old, old)` for each one. This gives a controller that has somehow missed its

> reconcile window (a stuck worker, a dropped handler) a chance to see the world again. It

Reviewer 2: Whats a "dropped handler"? And I don't think a `controller that has missed its reconcile window` is a thing, but they are frequently used to reconcile things outside the k8s api to which no event subscription exists (like resources in a cloudprovider), so things getting changed out of band without the controller getting an event for that is the typical reason for wanting resync

Author: Agreed, that was the wrong rationale. Reworded around out-of-band changes to external state, which is the actual reason resync exists.

### Mistake 3: resync is not relist

An informer has a resync period (`cache.Options.SyncPeriod`, 10 hours by default in
`controller-runtime`), and many people read it as "rebuild the cache from the API server every

> N hours".

Reviewer: An informer has a resync period (`cache.Options.SyncPeriod`, 10 hours by default in
`controller-runtime`), and many people read it as meaning: _rebuild the cache from the API server every_ n _hours, fetching every resource once again._
```

??

Author: Works for me, applied.

both cheaper and more correct than a hand-rolled timer: you do not hold a worker, and you do
not risk missing a real event.

> There is also `ctrl.Result{Requeue: true}`, which enqueues subject to the rate limiter, but it

Reviewer 2: Maybe just don't mention it, given its deprecated and shouldn't be used?

Author: Removed.

}
```

> It works — until the cluster has 50,000 Pods and reconciles run hundreds of times per second.

Reviewer 2: I don't think someone who can afford to have a 50k pod cluster cares the slightest about the memory cost of this, but they care about the fact that this control loop will be very slow due to heavy lock contention

Author: Reframed around latency and lock contention rather than allocations.

}
```

It works — until the cluster has 50,000 Pods and reconciles run hundreds of times per second.
Then the loop turns slow: every trigger walks all 50,000 Pods under the store's read lock,

> then deep-copies each one after the lock is released — `O(n)` work per reconcile, and it is

Reviewer: ```suggestion
then deep-copies each one after the lock is released, doing O(n) work per reconcile, and it is
```

Author: Applied.

string* comes back in `MatchingFields` at query time. Naming the index after the field it
happens to read is a readability convention, nothing more.

> That freedom holds only as long as the query stays in the cache. The same `MatchingFields` on

Reviewer 2: Maybe a less prosaic `But remember, this only works for reads against the cache` or such is enough rather than enumerating every way not to read from the cache?

Author: Replaced the paragraph with your one-liner.

an `APIReader`, on a type you listed in `Cache.DisableFor`,
or on an `Unstructured` object goes to the API server, which validates field names against the

> fields it actually indexes. `IndexField`'s own documentation says as much: "If you want

Reviewer 2: This second sentence is IMHO confusing as it seems to imply creating indexes that the apiserver doesn't have is a bad idea, which is absolutely not the case

Author: Gone along with the rest of the paragraph.

  value (for example, `now.Truncate(5*time.Minute).Format(...)`). You can then select objects
  by a specific window.
- **`MatchingLabels` is not an index.** Many people assume that since label-based lookups are
  so common, there must be an optimization for them. There is not: `ThreadSafeStore` keeps no
  separate label dictionary. The only indexes the cache has are the namespace index and the
  field indexes you register yourself.

>   Two separate things follow, and they get conflated. The walk really is `O(n)`: with no label

Reviewer: ```suggestion
  Two separate things follow, and they get conflated. The walk really is O(n): with no label
```

After we've done the important fixes, we could look at using math styling for O(n) etc. But, don't use backticks. It's not code.

Author: Dropped the backticks here and on the other O(n) mentions.

`mgr.GetAPIReader()` returns a `client.Reader` that goes straight to the API server, around
the cache. When you actually need it:

> - **Validating webhooks**, when the verdict depends on objects other than the one under

Reviewer 2: I have trouble understanding this whole webhook example. If the admission request runs in parallel with another request for another object that influences the admission decision, then this will spuriously race with or without apireader. Its totally possible etcd is in the middle of committing a parallel request and thats not something your apireader will see.

Furthermore, admission webhooks are typically run with more than one replica so if you are using them for quota admission its totally possible that two different instances of the webhook server get a request at approximately the same time and both admit it and thus allow to go over quota, the apireader isn't going to protect you from this at all. You need to explicitly increment a counter somewhere to guarantee correctness, at which point you can just rely on OCC and don't need an apireader.

This whole problem is basically why the eviction api grants evictions by updating the PDB status rather than trying to do some sort of consistent read because that just doesn't help if there is another apiserver. Quota admission plugin works the same IIRC.

IMHO bringing up the apireader in the webhook context is not useful and may confuse ppl into thinking their webhooks are race free just because they use an api reader.

Author: You're right, and this answers your original question better than the rewrite did: the webhook example shouldn't be there at all. A consistent-enough read does not exist under concurrent writers, and OCC is what actually guarantees correctness. Removed the webhook example everywhere and linked the FAQ instead.

alternative to `APIReader`: where `APIReader` is reached for ad hoc, individual requests,
`DisableFor` turns the cache off for the type wholesale.

> Real-world projects use this: [external-secrets](https://github.com/external-secrets/external-secrets),

Reviewer 2: Its worth mentioning that that by default also entails you are not getting events for relevant edges so you suddenly have to rely on periodically and I guess frequently re-reconciling in case the secret or configmap has changed, so its probably advisable to pair this with a metadata only watch

Author: Added, paired with the metadata-only pointer.

  gigabytes.
- **Do not call `mgr.GetClient()` before `mgr.Start()`.** The cache is not running yet, so
  reads fail with `ErrCacheNotStarted` instead of returning data. Use `mgr.GetAPIReader()`
  when you genuinely need to read during initialization.
- **For deferred actions, use `RequeueAfter`,** not `time.Sleep` and not your own goroutines.

## Wrapping up

In one breath:

- The cache in `controller-runtime` is not an optimization, it is the operating model. Under
  the hood it is `Reflector` + a delta queue + `Indexer` — exactly the same primitives that
  power Kubernetes itself.
- `r.Get` and `r.List` go to memory; `Create`, `Update`, `Patch`, and `Delete` go straight
  to the API server. Feedback flows in through the watch.
- `IndexField` plus `MatchingFields` turn the cache into a near-complete query engine with
  inverted indexes.
- `Namespaces`, selectors, `PartialObjectMetadata`, and `Transform` are the levers that
  control how much memory and traffic you actually consume.
- `APIReader` bypasses the cache for the rare read the cache cannot serve — but it is not a
  fix for staleness races; see the

>   [controller-runtime FAQ](https://github.com/kubernetes-sigs/controller-runtime/blob/main/FAQ.md#q-my-cache-might-be-stale-if-i-read-from-a-cache-how-should-i-deal-with-that).

Reviewer: nit: link to a stable ref, eg 820ed1a84f67c2f5c4ee4a104bc6c00faddb1384
```suggestion
- `APIReader` bypasses the cache for the rare read the cache cannot serve — but it is not a
  fix for staleness races; see the
  [controller-runtime FAQ](https://github.com/kubernetes-sigs/controller-runtime/blob/820ed1a84f67c2f5c4ee4a104bc6c00faddb1384/FAQ.md#q-my-cache-might-be-stale-if-i-read-from-a-cache-how-should-i-deal-with-that).
```
It may go stale, but so will the article (we automatically add a "this is stale" banner to old blog articles).

Author: Pinned.

  types: Secret, ConfigMap, Event, Pod, Node.
- **Add a `Transform`** for objects whose heavy fields you do not need —
  `ManagedFields` alone consume a noticeable share of memory.

> - **Add an `IndexField`** for every `List` that uses `MatchingFields`. Without a matching

Reviewer 2: I'd just drop the second sentence - People will notice this immediately the moment they run their controller

Author: True, the failure is loud, but only at runtime, the first time that particular List path is exercised. In the interest of readers not having to discover contracts by running into them, we'd like the checklist to be self-contained, so I'd prefer to keep the sentence.

  not the regular client.
- **Use `PartialObjectMetadata`** for types where you only need metadata. It can save
  gigabytes.

> - **Do not call `mgr.GetClient()` before `mgr.Start()`.** The cache is not running yet, so

Reviewer 2: I would not bother mentioning this here, just because the fact that it errors makes it impossible not to notice this is wrong.  I probably would mention that the cache acts as if objects not matching its selectors don't exist

Author: Added the selector-invisibility point as its own bullet next to "constrain cache scope". Kept the GetClient one: the error is loud, but the bullet also says what to reach for instead (APIReader) before you hit it, and we'd like the checklist to be self-contained.

  reads fail with `ErrCacheNotStarted` instead of returning data. Use `mgr.GetAPIReader()`
  when you genuinely need to read during initialization.
- **For deferred actions, use `RequeueAfter`,** not `time.Sleep` and not your own goroutines.

## Wrapping up

In one breath:

- The cache in `controller-runtime` is not an optimization, it is the operating model. Under
  the hood it is `Reflector` + a delta queue + `Indexer` — exactly the same primitives that
  power Kubernetes itself.
- `r.Get` and `r.List` go to memory; `Create`, `Update`, `Patch`, and `Delete` go straight
  to the API server. Feedback flows in through the watch.
- `IndexField` plus `MatchingFields` turn the cache into a near-complete query engine with
  inverted indexes.
- `Namespaces`, selectors, `PartialObjectMetadata`, and `Transform` are the levers that
  control how much memory and traffic you actually consume.

> - `APIReader` is the emergency exit when this process's cache lag is not acceptable.

Reviewer 2: This blog makes me want to deprecate the apireader lol. You should not use it, its not good enough to guarantee correctness because a concurrent request may be mid commit and you should find another way to deal with the cache lag. Can we instead link to the FAQ section on this maybe: https://github.com/kubernetes-sigs/controller-runtime/blob/main/FAQ.md#q-my-cache-might-be-stale-if-i-read-from-a-cache-how-should-i-deal-with-that

Author: Linked the FAQ here and in the read-after-write section.
